README.md
# Klipper-WS281x_LED_Status
This script will take the printer status from Klipper/Moonraker and apply different effects to a WS281x LED strip.

The code has been migrated from the OctoPrint-WS281x_LED_Status (https://github.com/cp2004/OctoPrint-WS281x_LED_Status) plugin to work with Klipper.

----

## Directions for use

1. Install prerequsits
   1. ```sudo apt update && sudo apt install -y git```
   2. ```sudo pip3 install requests PyYAML RPi.GPIO rpi_ws281x adafruit-circuitpython-neopixel```
2. Clone code to Raspberry Pi running Klipper and Moonraker
   1. ```cd /home/pi```
   2. ```git clone https://github.com/11chrisadams11/Klipper-WS281x_LED_Status.git```
   3. ```cd Klipper-WS281x_LED_Status```
3. Make script executable
   1. ```chmod 744 ./klipper_ledstrip.py```
4. Change strip values in settings.conf (LED pin, brightness, timeout)
5. Optionally, change effects and colors for standby, paused, and error states in settings.conf
6. If you want to run it manually, start script before starting print (otherwise use the service below)
   1. ```./klipper_ledstrip.py```

## Directions to run as a systemd service

1. Copy contents of ledstrip.service to /etc/systemd/system/ledstrip.service
2. Modify User, Group, WorkingDirectory, and ExecStart to match your setup
3. Run ```systemctl daemon-reload``` to enable the service
4. Run ```systemctl enable ledstrip``` to have the service start on boot
5. Run ```systemctl start ledstrip``` to start the service

## Directions to change settings (when using service)

1. Modify settings in settings.conf
2. Run ```systemctl restart ledstrip``` to restart the ledstrip service

### Single run for static colors
#### Will only work by itself, not if running as a service

```
./klipper_ledstrip.py <red> <green> <blue> <brightness:optiona>

Example:
  ./klipper_ledstrip.py 255 255 255 255 ## Full brightness white
  ./klipper_ledstrip.py 255 0 0 ## Red with default brightness specified in the script
```

#### To call from gcode shell commands (thanks to [JV_JV](https://www.reddit.com/user/JV_JV/) for the setup directions)
Add custom entries to printer.cfg 

```
[gcode_shell_command led_off]
command: ./home/pi/my_klipper_ledstrip.py 0 0 0
timeout: 2.
verbose: True

[gcode_shell_command led_white]
command: ./home/pi/my_klipper_ledstrip.py 255 255 255
timeout: 2.
verbose: True

[gcode_shell_command led_purple]
command: ./home/pi/my_klipper_ledstrip.py 255 0 255
timeout: 2.
verbose: True

[gcode_macro LED_OFF]
gcode:
    RUN_SHELL_COMMAND CMD=led_off

[gcode_macro LED_WHITE]
gcode:
    RUN_SHELL_COMMAND CMD=led_white

[gcode_macro LED_PURPLE]
gcode:
    RUN_SHELL_COMMAND CMD=led_purple
```

----

rpi_ws281x library instructions for needed changes depending on GPIO pin used: https://github.com/jgarff/rpi_ws281x


CHANGELOG.md
# Change Log

----

## 2023-03-04
Remove power off functions and add Moonraker API class

### Changed
- Removed settings and functions for powering off printer, as it was not related to controlling LED strips (moved to https://github.com/11chrisadams11/klipper-power-off-on-completion)
- Moved Moonraker API functions to a class
- General cleanup

----

## 2022-05-15
Threading and settings sample

### Changed
- Moved non-progress effects to their own thread to avoid pause between status checks
- Cleaned up effects class to work with threads
- Renamed settings.conf to settings_sample.conf to not overwrite previous settings
- Write sample settings if file not found
- Update service sample to clear LED strip on service stop

----

## 2022-04-04
More effects and a lot of cleanup.
The settings format has changed a bit, so you might need to redo your changes (Sorry)

### Added
- More effects (fill, fill_unfill, fill_chase, twinkle, twinkle_colors, noise, wave)
- Ability to specify a second color to blend with first
- Rainbow as a color choice

### Changed
- Moved effects to a class to enable effects to work with more color selections
- Cleaned up testing outputs
- Tried to make the install instruction more clear

### Fixed
- Fixed command in README.md (thanks [NatatorD](https://github.com/11chrisadams11/Klipper-WS281x_LED_Status/issues/9))

utils.py
'''
Utility functions for LED strip
'''

def average(num_a, num_b):
    ''' Average two given numbers '''
    return round((num_a + num_b) / 2)


def mix_color(colour1, colour2, percent_of_c1=None):
    ''' Mix two colors to a given percentage '''
    if percent_of_c1:
        colour1 = [x * percent_of_c1 for x in colour1]
        percent_of_c2 = 1 - percent_of_c1
        colour2 = [x * percent_of_c2 for x in colour2]

    col_r = average(colour1[0], colour2[0])
    col_g = average(colour1[1], colour2[1])
    col_b = average(colour1[2], colour2[2])
    return tuple([int(col_r), int(col_g), int(col_b)])


def color_brightness_correction(color, brightness):
    ''' Adjust given color to set brightness '''
    brightness_correction = brightness / 255
    return (
        int(color[0] * brightness_correction),
        int(color[1] * brightness_correction),
        int(color[2] * brightness_correction)
    )


moonraker_api.py
'''
Functions to connect to the Moonraker API and perform actions.
'''
import json
import math
import requests


class MoonrakerAPI:
    ''' Create Moonraker API class '''
    def __init__(self, moonraker_settings):
        self.moonraker_host = moonraker_settings['host']
        self.moonraker_port = str(moonraker_settings['port'])
        self.moonraker_url = f"http://{self.moonraker_host}:{self.moonraker_port}"
        self.bed_base_temp = False
        self.extruder_base_temp = False

    def printer_state(self):
        ''' Get printer status '''
        url = f"{self.moonraker_url}/printer/objects/query?print_stats"
        try:
            ret = requests.get(url)
        except requests.exceptions.ConnectionError:
            return False
        try:
            return ret.json()['result']['status']['print_stats']['state']
        except KeyError:
            return False

    def printing_stats(self):
        ''' Get stats for bed heater, hotend, and printing percent '''
        url = f"{self.moonraker_url}/printer/objects/query?heater_bed&extruder&display_status"
        data = requests.get(url).json()

        bed_temp = data['result']['status']['heater_bed']['temperature']
        bed_target = data['result']['status']['heater_bed']['target']
        ## Set base temperatures to make heating progress start from the bottom of strip
        if not self.bed_base_temp:
            self.bed_base_temp = bed_temp if bed_temp else 0

        extruder_temp = data['result']['status']['extruder']['temperature']
        extruder_target = data['result']['status']['extruder']['target']
        ## Set base temperatures to make heating progress start from the bottom of strip
        if not self.extruder_base_temp:
            self.extruder_base_temp = extruder_temp if extruder_temp else 0

        return {
            'bed': {
                'temp': float(bed_temp),
                'heating_percent': heating_percent(bed_temp, bed_target, self.bed_base_temp),
                'power_percent': round(data['result']['status']['heater_bed']['power'] * 100)
            },
            'extruder': {
                'temp': float(extruder_temp),
                'heating_percent': heating_percent(extruder_temp, extruder_target, self.extruder_base_temp),
                'power_percent': round(data['result']['status']['extruder']['power'] * 100)
            },
            'printing': {
                'done_percent': round(data['result']['status']['display_status']['progress'] * 100)
            }
        }

def heating_percent(temp, target, base_temp):
    ''' Get heating percent for given component '''
    if target == 0.0:
        return 0
    return math.floor(((temp - base_temp) * 100) / (target - base_temp))


klipper_ledstrip.py
#!/usr/bin/python3 -u
# pylint: disable=C0326,C0301
'''
Script to take info from Klipper and light up WS281x LED strip based on current status
'''
import os
import sys
import time
import yaml
import threading
from rpi_ws281x import Adafruit_NeoPixel
import moonraker_api
import effects
import utils

def get_settings():
    ''' Read settings from file '''
    script_path = os.path.dirname(os.path.realpath(__file__))
    try:
        with open(f'{script_path}/settings.conf', 'r') as settings_file:
            try:
                settings = yaml.safe_load(settings_file)
                return settings
            except yaml.scanner.ScannerError as err:
                print(f'\nSettings file formatted incorrectly:\n\t{err}')
                sys.exit()
    except FileNotFoundError:
        print('\nSettings file (settings.conf) not found. Adding sample settings.')
        with open(f'{script_path}/settings_sample.conf', 'r') as sample_settings_file:
            with open(f'{script_path}/settings.conf', 'w') as settings_file:
                settings_file.write(sample_settings_file.read())
        return get_settings()


def set_strip(strip_settings):
    strip = Adafruit_NeoPixel(
        strip_settings['led_count'],
        strip_settings['led_pin'],
        strip_settings['led_freq_hz'],
        strip_settings['led_dma'],
        strip_settings['led_invert'],
        strip_settings['led_brightness'],
        strip_settings['led_channel'],
    )
    return strip


def run():
    ''' Do work son '''
    settings = get_settings()
    
    moonraker_settings = settings['moonraker_settings']
    moonraker_api_cl = moonraker_api.MoonrakerAPI(moonraker_settings)
    
    strip_settings = settings['strip_settings']
    effects_settings = settings['effects']
    strip = set_strip(strip_settings)
    strip.begin()

    effects_cl = effects.Effects(strip, strip_settings, effects_settings)
    bed_progress = effects.Progress(strip, strip_settings, effects_settings['bed_heating'])
    hotend_progress = effects.Progress(strip, strip_settings, effects_settings['hotend_heating'])
    printing_progress = effects.Progress(strip, strip_settings, effects_settings['printing'])

    shutdown_counter = 0
    idle_timer = 0
    old_state = ''
    test_counter = 0
    try:
        while True:
            printer_state = moonraker_api_cl.printer_state()
            # print(printer_state)
            if printer_state == 'printing':
                printing_stats = moonraker_api_cl.printing_stats()
                printing_percent = printing_stats['printing']['done_percent']

                ## Set bed heating progress
                # print(printing_percent)
                if (printing_percent < 1 or printing_percent == 100) and printing_stats['bed']['heating_percent'] < 100:
                    bed_progress.set_progress(printing_stats['bed']['heating_percent'])

                ## Set hotend heating progress
                if (
                    (printing_percent < 1 or printing_percent == 100) and
                    printing_stats['extruder']['heating_percent'] < 100 and
                    printing_stats['bed']['heating_percent'] >= 99
                ):
                    hotend_progress.set_progress(printing_stats['extruder']['heating_percent'])

                ## Clear strip if bed and hotend heating are both done and print percent is 0
                if (
                    printing_percent == 0 and
                    printing_stats['extruder']['heating_percent'] >= 100 and
                    printing_stats['bed']['heating_percent'] >= 100
                ):
                    printing_progress.clear_strip()

                ## Set printing progress
                if 0 < printing_percent < 100:
                    printing_progress.set_progress(printing_percent)

            if printer_state != 'printing' and old_state == printer_state:
                idle_timer += 2
                if idle_timer > strip_settings['idle_timeout']:
                    effects_cl.stop_thread()
                    while effects_cl.effect_running:
                        time.sleep(0.1)
                    effects_cl.clear_strip()
            else:
                idle_timer = 0

            if old_state != printer_state:
                effects_cl.stop_thread()
                while effects_cl.effect_running:
                    time.sleep(0.1)
                effects_cl.start_thread()
                if printer_state in ['complete', 'standby', 'paused', 'error']:
                    effect_thread = threading.Thread(target=effects_cl.run_effect, args=(printer_state,)).start()

            old_state = printer_state
            time.sleep(2)

    except:
        effects_cl.stop_thread()
        while effects_cl.effect_running:
            time.sleep(0.1)
        effects_cl.clear_strip()

if __name__ == '__main__':
    if len(sys.argv) > 1:

        strip_settings = get_settings()['strip_settings']

        strip = set_strip(strip_settings)
        strip.begin()

        color = (int(sys.argv[1]), int(sys.argv[2]), int(sys.argv[3]))
        brightness = int(sys.argv[4]) if len(sys.argv) > 4 else strip_settings['led_brightness']

        for pixel in range(strip.numPixels()):
            strip.setPixelColorRGB(pixel, *utils.color_brightness_correction(color, brightness))
        strip.show()
    else:
        run()


effects.py
# pylint: disable=C0301
'''
LED strip effects functions
'''
import math
import time
from random import randint
import utils


class Effects:
    ''' Create effect class '''
    def __init__(self, strip, strip_settings, effects_settings):
        self.thread_stopped = False
        self.strip = strip
        self.strip_brightness = strip_settings['led_brightness']
        self.effects_settings = effects_settings
        self.printer_state = 'standby'
        self.effect_running = False
        self.effect_speed = ''
        self.effect_reverse = ''

        self.pixel_map = {
            'complete': [],
            'standby': [],
            'paused': [],
            'error': [],
        }
        for state in self.pixel_map:
            effect_color_1 = effects_settings[state]['color_1'] if ('color_1' in effects_settings[state] and effects_settings[state]['color_1']) else [255, 255, 255]
            effect_color_2 = effects_settings[state]['color_2'] if ('color_2' in effects_settings[state] and effects_settings[state]['color_2']) else None
            self.pixel_map[state] = self.set_pixel_map(effect_color_1, effect_color_2)

    def stop_thread(self):
        self.thread_stopped = True

    def start_thread(self):
        self.thread_stopped = False

    def set_pixel_map(self, effect_color_1, effect_color_2):
        ''' Create a list of pixel colors based on color settings and strip length '''
        pixel_map = []
        if effect_color_1 != 'rainbow' and not effect_color_2:
            for i in range(self.strip.numPixels()):
                pixel_map.append(effect_color_1)
        elif effect_color_1 != 'rainbow' and effect_color_2:
            spacing = 100 / (self.strip.numPixels() - 2)
            for i in range(self.strip.numPixels()):
                if i == 0:
                    pixel_map.append(effect_color_1)
                elif i == self.strip.numPixels() - 1:
                    pixel_map.append(effect_color_2)
                else:
                    pixel_map.append(utils.mix_color(effect_color_2, effect_color_1, (spacing * i) / 100))
        elif effect_color_1 == 'rainbow':
            colors = {
                'violet': (148, 0  , 211),
                'purple': (75 , 0  , 130),
                'blue'  : (0  , 0  , 255),
                'green' : (0  , 255, 0  ),
                'yellow': (255, 255, 0  ),
                'orange': (255, 127, 0  ),
                'red'   : (255, 0  , 0  )
            }
            halfish_pixels = math.floor(self.strip.numPixels() / 2)
            bottom_half_spacing = 100 / (halfish_pixels - 1)
            top_half_spacing = 100 / ((self.strip.numPixels() - halfish_pixels))
            upper_count = 1

            for i in range(self.strip.numPixels()):
                if i == 0:
                    pixel_map.append(colors['red'])
                if 0 < i < (halfish_pixels - 1):
                    pixel_map.append(utils.mix_color(colors['green'], colors['red'], (bottom_half_spacing * i) / 100))
                if i == (halfish_pixels - 1):
                    pixel_map.append(colors['green'])
                if (halfish_pixels - 1) < i < self.strip.numPixels():
                    pixel_map.append(utils.mix_color(colors['purple'], colors['green'], (top_half_spacing * upper_count) / 100))
                    upper_count += 1

        return pixel_map
        

    def get_pixel_map(self):
        ''' Return created pixel map '''
        return self.pixel_map

    def set_speed(self, slow, fast):
        ''' Set speed factor based on given setting '''
        if isinstance(self.effect_speed, (int, float)):
            speed = self.effect_speed
        elif self.effect_speed.lower() not in ['fast', 'slow']:
            speed = fast
        else:
            speed = slow if self.effect_speed.lower() == 'slow' else fast
        return speed

    def run_effect(self, printer_state):
        ''' Run the effect specified '''
        self.printer_state = printer_state
        effect = self.effects_settings[printer_state]['effect'] if 'effect' in self.effects_settings[printer_state] else 'solid'
        if effect not in ['solid', 'fade', 'chase', 'bounce', 'chase_ghost', 'ghost_bounce', 'fill', 'fill_unfill', 'fill_chase', 'twinkle', 'twinkle_colors', 'noise', 'wave', 'slava_ukraini']:
            effect = 'solid'
        self.effect_speed = self.effects_settings[printer_state]['speed'] if 'speed' in self.effects_settings[printer_state] else 'fast'
        self.effect_reverse = self.effects_settings[printer_state]['reverse'] if 'reverse' in self.effects_settings[printer_state] else False

        while not self.thread_stopped:
            self.effect_running = True
            eval(f"self.{effect}()")
            self.effect_running = False

    def clear_strip(self):
        ''' Turn all pixels of LED strip off '''
        for i in range(self.strip.numPixels()):
            self.strip.setPixelColorRGB(i, 0, 0, 0)
        self.strip.show()

    def solid(self):
        ''' Set static color for entire strip with no effect '''
        for pixel, color in enumerate(self.pixel_map[self.printer_state]):
            self.strip.setPixelColorRGB(pixel, *utils.color_brightness_correction(color, self.strip_brightness))
        self.strip.show()

    def fade(self):
        ''' Fade entire strip with given color and speed '''
        speed = self.set_speed(0.01, 0.005)
        for pixel, color in enumerate(self.pixel_map[self.printer_state]):
            self.strip.setPixelColorRGB(pixel, *color)
        self.strip.show()

        for i in range(self.strip_brightness):
            self.strip.setBrightness(i)
            self.strip.show()
            time.sleep(speed)

        time.sleep(speed * 5)

        for i in range(self.strip_brightness, -1, -1):
            self.strip.setBrightness(i)
            self.strip.show()
            time.sleep(speed)
        
        time.sleep(speed * 5)

    def chase(self):
        ''' Light one LED from one ond of the strip to the other, optionally reversed '''
        speed = self.set_speed(0.01, 0.005)
        self.strip.setBrightness(self.strip_brightness)
        for i in reversed(range(len(self.pixel_map[self.printer_state]) + 1)) if self.effect_reverse else range(len(self.pixel_map[self.printer_state]) + 1):
            for pixel, color in enumerate(self.pixel_map[self.printer_state]):
                if i == pixel:
                    self.strip.setPixelColorRGB(
                        pixel,
                        *utils.color_brightness_correction(
                            color,
                            self.strip_brightness
                        )
                    )
                else:
                    self.strip.setPixelColorRGB(pixel, 0, 0, 0)
                self.strip.show()
                time.sleep(speed)
        if self.effect_reverse:
            self.clear_strip()

    def bounce(self):
        ''' Bounce one LED back and forth '''
        self.chase()
        self.effect_reverse = not self.effect_reverse
        self.chase()
        self.effect_reverse = not self.effect_reverse

    def chase_ghost(self):
        ''' Light one LED from one ond of the strip to the other, optionally reversed '''
        speed = self.set_speed(0.01, 0.005)
        self.strip.setBrightness(self.strip_brightness)
        for i in reversed(range(len(self.pixel_map[self.printer_state]) + 5)) if self.effect_reverse else range(len(self.pixel_map[self.printer_state]) + 5):
            for pixel, color in enumerate(self.pixel_map[self.printer_state]):
                if i == pixel:
                    brightness = self.strip_brightness / 4 if self.effect_reverse else self.strip_brightness
                    self.strip.setPixelColorRGB(
                        pixel,
                        *utils.color_brightness_correction(
                            color,
                            brightness
                        )
                    )
                elif i - 1 == pixel:
                    brightness = (self.strip_brightness / 4) * 2 if self.effect_reverse else (self.strip_brightness / 4) * 3
                    self.strip.setPixelColorRGB(
                        pixel,
                        *utils.color_brightness_correction(
                            color,
                            brightness
                        )
                    )
                elif i - 2 == pixel:
                    brightness = (self.strip_brightness / 4) * 3 if self.effect_reverse else (self.strip_brightness / 4) * 2
                    self.strip.setPixelColorRGB(
                        pixel,
                        *utils.color_brightness_correction(
                            color,
                            brightness
                        )
                    )
                elif i - 3 == pixel:
                    brightness = self.strip_brightness if self.effect_reverse else self.strip_brightness / 4
                    self.strip.setPixelColorRGB(
                        pixel,
                        *utils.color_brightness_correction(
                            color,
                            brightness
                        )
                    )
                else:
                    self.strip.setPixelColorRGB(pixel, 0, 0, 0)
                self.strip.show()
                time.sleep(speed)
        if self.effect_reverse:
            self.clear_strip()

    def ghost_bounce(self):
        ''' Bounce one LED back and forth '''
        self.chase_ghost()
        self.effect_reverse = not self.effect_reverse
        self.chase_ghost()
        self.effect_reverse = not self.effect_reverse

    def fill(self, delay=True):
        ''' Fill strip one pixel at a time '''
        speed = self.set_speed(0.1, 0.05)
        self.clear_strip()
        for pixel in reversed(range(len(self.pixel_map[self.printer_state]))) if self.effect_reverse else range(len(self.pixel_map[self.printer_state])):
            self.strip.setPixelColorRGB(pixel, *self.pixel_map[self.printer_state][pixel])
            self.strip.show()
            time.sleep(speed)
        if delay:
            time.sleep(speed * 5)

    def fill_unfill(self):
        ''' Fill strip one pixel at a time and clear in reverse '''
        speed = self.set_speed(0.1, 0.05)
        self.fill(delay=False)
        time.sleep(speed * 2)
        for pixel in reversed(range(len(self.pixel_map[self.printer_state]), -1, -1)) if self.effect_reverse else range(len(self.pixel_map[self.printer_state]), -1, -1):
            self.strip.setPixelColorRGB(pixel, 0, 0, 0)
            self.strip.show()
            time.sleep(speed)
        time.sleep(speed * 2)

    def fill_chase(self):
        ''' Fill strip one pixel at a time and clear in chase '''
        speed = self.set_speed(0.1, 0.05)
        self.fill(delay=False)
        time.sleep(speed * 2)
        for pixel in reversed(range(len(self.pixel_map[self.printer_state]))) if self.effect_reverse else range(len(self.pixel_map[self.printer_state])):
            self.strip.setPixelColorRGB(pixel, 0, 0, 0)
            self.strip.show()
            time.sleep(speed)

    def twinkle(self):
        ''' Flash single pixels, in specified color(s), at random '''
        speed = self.set_speed(0.1, 0.05)
        self.strip.setBrightness(self.strip_brightness)
        self.clear_strip()
        for _ in range(int(2 / speed)):
            pixel = randint(0, len(self.pixel_map[self.printer_state]) - 1)
            self.strip.setPixelColorRGB(pixel, *self.pixel_map[self.printer_state][pixel])
            self.strip.show()
            time.sleep(speed)
            self.clear_strip()

    def twinkle_colors(self):
        ''' Flash single pixels, in random colors, at random '''
        speed = self.set_speed(0.1, 0.05)
        self.strip.setBrightness(self.strip_brightness)
        self.clear_strip()
        for r in range(int(2 / speed)):
            i = randint(0, self.strip.numPixels() - 1)
            r = randint(0, 255)
            g = randint(0, 255)
            b = randint(0, 255)
            self.strip.setPixelColorRGB(i, r, g, b)
            self.strip.show()
            time.sleep(speed)
            self.clear_strip()

    def noise(self):
        ''' Flash multiple pixels, in random colors, at random '''
        speed = self.set_speed(0.1, 0.05)
        self.strip.setBrightness(self.strip_brightness)
        for r in range(int(2 / speed)):
            for _ in range(randint(1, self.strip.numPixels() / 2)):
                rand_off = randint(0, 1)
                i = randint(0, self.strip.numPixels() - 1)
                r = randint(0, 255) if rand_off else 0
                g = randint(0, 255) if rand_off else 0
                b = randint(0, 255) if rand_off else 0
                self.strip.setPixelColorRGB(i, r, g, b)
            self.strip.show()
            time.sleep(speed)

    def wave(self):
        ''' Simulate waving flag '''
        speed = self.set_speed(0.1, 0.05)
        self.strip.setBrightness(self.strip_brightness)

        for i in reversed(range(len(self.pixel_map[self.printer_state]) + 8)) if self.effect_reverse else range(len(self.pixel_map[self.printer_state]) + 8):
            for pixel, color in enumerate(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(pixel, *color)
            if 0 <= i < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i, *utils.color_brightness_correction(self.pixel_map[self.printer_state][i], 80))
            if 0 <= i - 1 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 1, *utils.color_brightness_correction(self.pixel_map[self.printer_state][i - 1], 60))
            if 0 <= i - 2 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 2, *utils.color_brightness_correction(self.pixel_map[self.printer_state][i - 2], 40))
            if 0 <= i - 3 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 3, *utils.color_brightness_correction(self.pixel_map[self.printer_state][i - 3], 20))
            if 0 <= i - 4 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 4, *utils.color_brightness_correction(self.pixel_map[self.printer_state][i - 4], 40))
            if 0 <= i - 5 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 5, *utils.color_brightness_correction(self.pixel_map[self.printer_state][i - 5], 60))
            if 0 <= i - 6 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 6, *utils.color_brightness_correction(self.pixel_map[self.printer_state][i - 6], 80))
            self.strip.show()
            time.sleep(speed)

    def slava_ukraini(self):
        ''' Simulate waving flag '''
        speed = self.set_speed(0.1, 0.05)
        self.strip.setBrightness(self.strip_brightness)
        color1 = [0, 0, 255] if self.effect_reverse else [255, 255, 0]
        color2 = [255, 255, 0] if self.effect_reverse else [0, 0, 255]

        for i in reversed(range(len(self.pixel_map[self.printer_state]) + 8)) if self.effect_reverse else range(len(self.pixel_map[self.printer_state]) + 8):
            for pixel in range(len(self.pixel_map[self.printer_state])):
                self.strip.setPixelColorRGB(pixel, *color1 if pixel < ((self.strip.numPixels() - 1) / 2) else color2)
            if 0 <= i < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i, *utils.color_brightness_correction(color1 if i < ((self.strip.numPixels() - 1) / 2) else color2, 80))
            if 0 <= i - 1 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 1, *utils.color_brightness_correction(color1 if i - 1 < ((self.strip.numPixels() - 1) / 2) else color2, 60))
            if 0 <= i - 2 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 2, *utils.color_brightness_correction(color1 if i - 2 < ((self.strip.numPixels() - 1) / 2) else color2, 40))
            if 0 <= i - 3 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 3, *utils.color_brightness_correction(color1 if i - 3 < ((self.strip.numPixels() - 1) / 2) else color2, 20))
            if 0 <= i - 4 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 4, *utils.color_brightness_correction(color1 if i - 4 < ((self.strip.numPixels() - 1) / 2) else color2, 40))
            if 0 <= i - 5 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 5, *utils.color_brightness_correction(color1 if i - 5 < ((self.strip.numPixels() - 1) / 2) else color2, 60))
            if 0 <= i - 6 < len(self.pixel_map[self.printer_state]):
                self.strip.setPixelColorRGB(i - 6, *utils.color_brightness_correction(color1 if i - 6 < ((self.strip.numPixels() - 1) / 2) else color2, 80))
            self.strip.show()
            time.sleep(speed)


class Progress:
    def __init__(self, strip, strip_settings, effect_settings):
        self.strip = strip
        self.strip_brightness = strip_settings['led_brightness']
        self.base_color = effect_settings['base_color']
        self.progress_color = effect_settings['progress_color']
        self.effect_reverse = effect_settings['reverse'] if 'reverse' in effect_settings else False
        self.num_pixels = self.strip.numPixels()

    def set_progress(self, percent):
        upper_bar = (percent / 100) * self.num_pixels
        upper_remainder, upper_whole = math.modf(upper_bar)
        pixels_remaining = self.num_pixels
        self.strip.setBrightness(self.strip_brightness)

        for i in range(int(upper_whole)):
            pixel = ((self.num_pixels - 1) - i) if self.effect_reverse else i
            self.strip.setPixelColorRGB(
                pixel,
                *utils.color_brightness_correction(
                    self.progress_color,
                    self.strip_brightness
                )
            )
            pixels_remaining -= 1

        if upper_remainder > 0.0:
            tween_color = utils.mix_color(self.progress_color, self.base_color, upper_remainder)
            pixel = (
                ((self.num_pixels - int(upper_whole)) - 1)
                if self.effect_reverse
                else int(upper_whole)
            )
            self.strip.setPixelColorRGB(
                pixel,
                *utils.color_brightness_correction(
                    tween_color,
                    self.strip_brightness
                )
            )
            pixels_remaining -= 1

        for i in range(pixels_remaining):
            pixel = (
                ((pixels_remaining - 1) - i)
                if self.effect_reverse
                else ((self.num_pixels - pixels_remaining) + i)
            )
            self.strip.setPixelColorRGB(
                pixel,
                *utils.color_brightness_correction(
                    self.base_color,
                    self.strip_brightness
                )
            )

        self.strip.show()

    def clear_strip(self):
        ''' Turn all pixels of LED strip off '''
        for i in range(self.strip.numPixels()):
            self.strip.setPixelColorRGB(i, 0, 0, 0)
        self.strip.show()
